File: merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from queue import PriorityQueue
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        dummy_head = tail = ListNode()
        
        pq = PriorityQueue()
        # dummy number to avoid ListNode comparison
        dummy_num = 0
        for head in lists:
            if head:
                pq.put((head.val, dummy_num, head))
                dummy_num += 1
        
        while not pq.empty():
            _, _, top = pq.get()
            if top.next:
                pq.put((top.next.val, dummy_num, top.next))
                dummy_num += 1
            tail.next = top
            tail = tail.next
        return dummy_head.next

File: merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        return merge(lists, 0, lists.size() - 1);
    }
private:
    ListNode* merge(vector<ListNode*>& lists, int l, int r){
        if (l > r) return nullptr;
        if (l == r) return lists[l];
        if (l + 1 == r) return mergeTwoLists(lists[l], lists[r]);
        
        int mid = l + (r - l) / 2;
        auto l1 = merge(lists, l, mid);
        auto l2 = merge(lists, mid + 1, r);
        return mergeTwoLists(l1, l2);
    }
    
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2){
        ListNode dummy_head(0);
        ListNode* tail = &dummy_head;
        
        while (l1 && l2){
            if (l1->val > l2->val) swap(l1, l2);
            tail->next = l1;
            l1 = l1->next;
            tail = tail->next;
        }
        tail->next = l1 ? l1 : l2;
        return dummy_head.next;
    }
};
